### PR TITLE
Move control byte to separate array. Add functions to modify dynamically

### PR DESCRIPTION
**Pull request by canardos**
Per issue #821, control flags have been moved from the map to a separate array for the button matrix and functions to modify the flags have been implemented.

### CODE REVIEW

**Comment by kisvegabor** [comment]
Looks good, thank you very much!

I'm missing only one thing. It should be a `lv_btnm_set_ctrl_map(btnm, ctrl_map)` function. It would make possible to declare the whole layout in compile time. For example, it is used by the [keyboard object ](https://github.com/littlevgl/lvgl/blob/master/lv_objx/lv_kb.c#L35).

**Comment by canardos** [comment]
That seems like a  good step on the road to removing the control chars from the string map.

The control bits can still currently be defined at compile time as they always have by using the control chars, but I agree that having a `lv_btnm_set_ctrl_map(btnm, ctrl_map)` function helps move things towards removing them completely.

One potential issue is that the current `lv_btnm_set_map` function was built with the expectation of knowing the control bits when it is called (since they are in the map). Specifically the relative widths.

Calling `lv_btnm_set_map` and then calling `lv_btnm_set_ctrl_map` necessitates calling  `lv_btnm_set_map` again (presumably from `lv_btnm_set_ctrl_map`). `lv_btnm_set_map` is also currently called from `lv_btnm_create` with the default map.

```
lv_btnm_create
    lv_btnm_set_map // with default map
lv_btnm_set_map // with new map
lv_btnm_set_ctrl_map
    lv_btnm_set_map // with new map
```

So `lv_btnm_set_map` is called a minimum of 3 times, with the full layout performed each time. I'm not sure exactly how expensive it is, but it seems a bit wasteful.

I'm not sure I have any solid ideas on how to avoid this with the code as it currently stands, although I would definitely suggest suggest changing `lv_btnm_def_map` to perhaps a single button to reduce the amount of time wasted laying out a map that won't be used.

A `lv_btnm_set_map(btnm, map, ctrl_map)` function would avoid the second call and is closer to the original idea of having the control bits in the map, but I'm not sure if it's in line with the current API which seems to restrict all set functions to setting a single parameter.

So I guess the choice is `lv_btnm_set_ctrl_map(btnm, ctrl_map)`, `lv_btnm_set_map(btnm, map, ctrl_map)`, or both.

**Comment by embeddedt** [comment]
@canardos 

>  I would definitely suggest changing `lv_btnm_def_map` to perhaps a single button to reduce the amount of time wasted laying out a map that won't be used.

I think one good reason for having a default map is so that the user can get a feel for what the widget will look like on their screen. It also allows them to quickly experiment with styles, etc. without having to create a button map first.

**Comment by canardos** [comment]
I agree that having a default map seems fine the reasons you mentioned. I was just suggesting that loading, parsing, allocating memory for it etc. prior to loading/parsing the real map for every control created seems a bit wasteful.

I would have expected it to be defined in `lv_btnm.h` to be used like `lv_btnm_set_map(btnm, lv_btnm_sample_map)` for example.

**Comment by embeddedt** [comment]
> I agree that having a default map seems fine the reasons you mentioned. I was just suggesting that loading, parsing, allocating memory for it etc. prior to loading/parsing the real map for every control created seems a bit wasteful.
> 
> I would have expected it to be defined in `lv_btnm.h` to be used like `lv_btnm_set_map(btnm, lv_btnm_sample_map)` for example.

Fair enough. The only issue with that is that the user would have to find the sample map themselves, instead of it being automatically set up for them.

**Comment by kisvegabor** [comment]
@canardos Applying the map several times is really wasteful and can have cause some performance loss. However, nobody complained about "slow object creation" so far so I think we can leave the default map. As @embeddedt has highlighted it's useful for the users to see how the object looks like without doing anything.


With your update, we have a nice API to control the buttons one-by-one. I assume most of the people will use it because it's very clear and readable. So the last parameter of `lv_btnm_set_map(btnm, map, ctrl_map)` might be used rarely. Therefore I vote for having a `lv_btnm_set_ctrl_map(btnm, ctrl_map)` which can be used in some special cases if required.

**Comment by canardos** [comment]
@kisvegabor No problem. I'll implement `lv_btnm_set_ctrl_map(btnm, ctrl_map)` in the next few days and update the PR.

**Comment by canardos** [comment]
Note that there are still a couple of issues to resolve post the above commits. Will let you know when it's ready for review.

**Comment by canardos** [comment]
@kisvegabor I think that should be all good. Please take a look.

**Review by kisvegabor — COMMENTED** [review]
@canardos I'm not sure you were notified about my comment on the code without explicitly mentioning you.

File: lv_objx/lv_btnm.c [hunk 1]
```diff
@@ -235,7 +235,7 @@ void lv_btnm_set_map(const lv_obj_t * btnm, const char ** map)
                 /* one_unit_w = all_unit_w / unit_cnt
                  * act_unit_w = one_unit_w * button_width
                  * do this two operations but the multiply first to divide a greater number */
-                act_unit_w = (all_unit_w * get_button_width(ext->ctrl_bits[i])) / unit_cnt;
+                act_unit_w = (all_unit_w * get_button_width(ext->ctrl_bits[btn_i])) / unit_cnt;
```
[the remark this answers is not in the document]
Comment (canardos) [reply lv_objx/lv_btnm.c:238]: Sorry, I didn't see the notification. Yes, that's correct. The length and indices of the `ext->ctrl_bits` array equals that of the `ext->button_areas` array. There are no elements for `\n` entries.

**Comment by kisvegabor** [comment]
As there is a plan to update `lv_btm` (https://github.com/littlevgl/lvgl/issues/901) I merge it now.

@canardos 
Please clarify the comment above.

Thank you very much for this update! It made `lv_btnm` more flexible and will help a lot to add new features.

**Comment by canardos** [comment]
> As there is a plan to update `lv_btm` (#901) I merge it now.
> 
> @canardos
> Please clarify the comment above.
> 
> Thank you very much for this update! It made `lv_btnm` more flexible and will help a lot to add new features.


@kisvegabor Thanks for reviewing and merging. Which comment did you want clarified?

**Comment by kisvegabor** [comment]
@canardos 

I commented on the commit but it seems people don't get notified if someone comments on their commit. Next time  I'll use the "mentioning".

So I ment:
Is this comment correct? 
https://github.com/littlevgl/lvgl/blob/688cd79a550e058aee0c9ad95cf5043bb7575887/lv_objx/lv_btnm.c#L272-L287

It says "one element for each button AND new line".

**Comment by canardos** [comment]
@kisvegabor  Ah, missed that one. Per my earlier reply, the array size/index is equal to that of the buttons and there are no entries for newlines. You'll see in commit https://github.com/littlevgl/lvgl/pull/825/commits/688cd79a550e058aee0c9ad95cf5043bb7575887 that I updated the comments in all of the other functions to reflect this, but it seems I missed that one. It should read:

```
  * @param ctrl_map pointer to an array of `lv_btn_ctrl_t` control bytes. The 
  *                 length of the array and position of the elements must match 
  *                 the number and order of the individual buttons (i.e. excludes
  *                 newline entries).
  *                 The control bits are: 
  *                 - bit 5   : 1 = inactive (disabled) 
  *                 - bit 4   : 1 = no repeat (on long press) 
  *                 - bit 3   : 1 = hidden 
  *                 - bit 2..0: Relative width compared to the buttons in the 
  *                             same row. [1..7]
```

**Comment by kisvegabor** [comment]
Thank you! I'll update it accordingly.

**Pull request merged**